src/PrestaShopBundle/Controller/Admin/Common/EntityController.php

```diff
@@ -0,0 +1,83 @@
+<?php
+/**
+ * 2007-2018 PrestaShop
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Controller\Admin\Common;
+
+use PrestaShop\PrestaShop\Core\Form\Entity\FormFactory\EntityFormFactoryInterface;
+use PrestaShop\PrestaShop\Core\Form\Entity\ResponseHandler\EntryResponseHandlerInterface;
+use PrestaShop\PrestaShop\Core\Form\Entity\ResponseHandler\FailureResponseHandlerInterface;
+use PrestaShop\PrestaShop\Core\Form\Entity\ResponseHandler\SuccessResponseHandlerInterface;
+use PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController;
```

Comment:
I like this idea of having a common controller and strongly typed Responses, I insist on the STRONGLY typed and here is why (but you probably had this in mind)
Having a common controller allows us to simply create new controller only through configuration and dependency injection, we could create a full controller with yaml only, based on this https://symfony.com/doc/3.4/controller.html#fetching-services-as-controller-arguments

```
# app/config/services.yml
services:
        core.prestashop.controller.product:
        class: PrestaShopBundle\Controller\Admin\Common\EntityController
        public: true
        bind:
            $addEntryResponseHandler: PrestaShop\PrestaShop\Core\Form\Entity\ResponseHandler\Common\ RedirectWithSuccessAddMessageHandler

    core.prestashop.controller.product:
        class: PrestaShopBundle\Controller\Admin\Common\EntityController
        public: true
        bind:
            $addEntryResponseHandler: '@core.prestashop.controller.product.add_success_handler'

    core.prestashop.controller.product.add_success_handler:
        class: PrestaShop\PrestaShop\Core\Form\Entity\ResponseHandler\Product\AddProductSuccessResponseHandler
        arguments:
            $cache: "@cache"
```

Where `AddProductSuccessResponseHandler` could be an extension of `RedirectWithSuccessAddMessageHandler` but it needs to clear the cache for a specific product for example (although it is a bad example because this cache flush should rather be done in a hook so that it works even outside of a controller context)
But it could be you want to change the output of your controller to transform it into a Rest controller to build an API

Anyway this offers lots of possibilities for creating controller quickly, as well as limiting the quantity of code and rather use some configuration, it also has a lot of possibilities concerning the overriding

Reply:
I am worried about a configurable common Controller because:
- it makes debugging hard (if you want to put a var_dump somewhere)
- everytime there will be a small custom behavior it will be extended anyway (for example in a CRUD page if there is a small info such as a warning message)
- it makes the code harder to understand as we have to keep it very generic with functions such as `handleEntity($entityName);`, `getIdentifierForEntity($entityName);` thus harder to understand, debug, extend, improve
- it makes prestashop harder to understand as people are used to the paradigm `1 route <-> 1 controller <-> 1 action <-> 1 template`; for example sometimes I'm on a front-office page and in order to find the controller used I copy-paste some html code then search this using my IDE to find the right file

Reply:
If we do not use a common EntityController, we can remove `EntryResponseHandlerInterface`, `FailureResponseHandlerInterface`, `EntityFormFactoryResolver` ... it makes the entity forms system easier.

We can use an AbstractEntityController to reuse code, we can use also some traits ... but having 1 controller and 1 action easily identifiable really makes the prestashop system easier.